consensus/src/execution_ctx.rs

```diff
@@ -149,79 +153,108 @@ impl<'a, DB: Database, T: Operations + 'static> ExecutionCtx<'a, DB, T> {
+                    self.executor.clone(),
+                )
+                .await;
+            };
+        } else {
+            error!(event = "committee not found", msg_iteration);
         }
     }
 
-    fn try_vote(&mut self, msg_step: u16, candidate: &Block, topic: Topics) {
-        if let Some(committee) =
-            self.iter_ctx.committees.get_committee(msg_step)
-        {
-            if self.am_member(committee) {
-                debug!(
-                    event = "vote for former candidate",
-                    step_topic = format!("{:?}", topic),
-                    hash = to_str(&candidate.header().hash),
-                    msg_step,
-                );
+    pub(crate) async fn try_cast_ratification_vote(
+        &self,
+        msg_iteration: u8,
+        validation: &ValidationResult,
+    ) {
+        let step = StepName::Ratification.to_step(msg_iteration);
 
-                // TODO: Verify
-            };
-        } else {
-            error!(event = "committee not found", step = self.step(), msg_step);
+        if let Some(committee) = self.iter_ctx.committees.get_committee(step) {
+            if self.am_member(committee) {
+                RatificationStep::<T, DB>::try_vote(
+                    &self.round_update,
+                    msg_iteration,
+                    validation,
+                    self.outbound.clone(),
+                )
+                .await;
+            }
         }
     }
 
     /// Process messages from past
     async fn process_past_events(&mut self, msg: Message) -> Option<Message> {
-        if msg.header.round != self.round_update.round {
+        if msg.header.round != self.round_update.round
+            || self.iteration < EMERGENCY_MODE_ITERATION_THRESHOLD
+        {
+            // Discard messages from past if current iteration is not considered
+            // an emergency iteration
             return None;
         }
 
+        self.on_emergency_mode(msg).await
+    }
+
+    /// Handles a consensus message in emergency mode
+    async fn on_emergency_mode(&mut self, msg: Message) -> Option<Message> {
         if let Err(e) = self.outbound.send(msg.clone()).await {
             error!("could not send msg due to {:?}", e);
         }
 
-        // Try to vote for candidate block from former iteration
+        // Try to cast validation vote for a candidate block from former
+        // iteration
         if let Payload::Candidate(p) = &msg.payload {
-            // TODO: Perform block header/ Certificate full verification
-            // To be addressed with another PR
-
-            self.vote_for_former_candidate(msg.header.get_step(), &p.candidate)
+            self.try_cast_validation_vote(msg.header.iteration, &p.candidate)
                 .await;
         }
 
-        // Collect message from a previous reduction step/iteration.
+        let msg_iteration = msg.header.iteration;
+
+        // Collect message from a previous iteration/step.
```

Comment:
I think this is just previous iteration

Reply:
A past event can be either a message from previous iteration or message from same iteration but previous step number. 
E.g while running iteration 0, Validation step - my node may receive the Candidate of current iteration.

That's why I put here `iteration/step`

Reply:
Yeah, I see that's coherent with the current implementation.
But, as mentioned in my other comment, in emergency mode we should not move to the next step until the previous one ended.
So here we would simply collect messages from previous iterations, (regardless of the current step).
Instead, wrt the current iteration, previous-step messages should be discarded